src/fm_block_allocator.c

```diff
@@ -0,0 +1,170 @@
+/****************************************************************************
+	[Project] FlexSEA: Flexible & Scalable Electronics Architecture
+	[Sub-project] 'flexsea-manage' Mid-level computing, and networking
+	Copyright (C) 2016 Dephy, Inc. <http://dephy.com/>
+
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*****************************************************************************
+	[Lead developper] Jean-Francois (JF) Duval, jfduval at dephy dot com.
+	[Origin] Based on Jean-Francois Duval's work at the MIT Media Lab
+	Biomechatronics research group <http://biomech.media.mit.edu/>
+	[Contributors]
+*****************************************************************************
+	[This file] fm_block_allocator: Fixed sized block allocator
+*****************************************************************************
+	[Change log] (Convention: YYYY-MM-DD | author | comment)
+	* 2017-01-20 | igutekunst | Initial implementation
+	*
+*****************************************************************************/
+#include "../inc/fm_block_allocator.h"
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+struct Block {
+	struct Block * next;
+	struct Block * prev;
+	size_t index;
+	char data[FM_BLOCK_SIZE];
+};
+
+
+typedef struct Block Block;
+
+Block memory_pool[FM_NUM_BLOCKS];
+
+#define offsetof(st, m) ((size_t)(&((st *)0)->m))
+
+static Block* get_block(void * raw_block) {
+	size_t offset = offsetof(Block, data);
+	Block* b = raw_block - offset;
```

Comment:
void pointer arithmetic is illegal in c / c++. Granted it will work with gcc, but maybe better to use char pointers?

http://stackoverflow.com/questions/3523145/pointer-arithmetic-for-void-pointer-in-c